Having more than two types param in a type and deriving Reflect errors

## Bevy version
0.15.3

## What you did

I get from this code snippet 
```Rust
use super::ScriptFunctionChecker;
use bevy_ecs::system::Resource;
use bevy_mod_scripting::core::event::IntoCallbackLabel;
use bevy_mod_scripting::core::script::ScriptId;
use bevy_mod_scripting::core::IntoScriptPluginParams;
use bevy_reflect::Reflect;
use indexmap::IndexSet;
use std::marker::PhantomData;

/// Type-parameterized script storage for tracking active scripts
#[derive(Resource, Default, Reflect)]
pub struct KomoToolScriptStore<P, L>
where
    P: IntoScriptPluginParams
        + ScriptFunctionChecker
        + Send
        + Sync
        + 'static
        + std::default::Default,
    L: IntoCallbackLabel + Send + Sync + 'static + std::default::Default,
{
    /// Set of active script identifiers
    #[reflect(ignore)]
    pub scripts: IndexSet<ScriptId>,
    #[reflect(ignore)]
    _phantom: PhantomData<(L, P)>,
}
```

## What went wrong
I get this error:

```Rust
error[E0277]: cannot add `&std::string::String` to `std::string::String`               
  --> crates\komotool_utils\src\handler\script_store.rs:11:29
   |
11 | #[derive(Resource, Default, Reflect)]
   |                             ^^^^^^^ no implementation for `std::string::String + &std::string::String`
   |
   = help: the trait `std::ops::Add<&std::string::String>` is not implemented for `std::string::String`
   = help: the following other types implement trait `std::ops::Add<Rhs>`:
             `std::string::String` implements `std::ops::Add<&str>`
             `std::string::String` implements `std::ops::Add<smartstring::SmartString<Mode>>`
   = note: this error originates in the derive macro `Reflect` (in Nightly builds, run with -Z macro-backtrace for more info)
```

## Additional information
I tried some things now to this code is also not compiling here is nothing use in BMS:
```Rust
use super::ScriptFunctionChecker;
use bevy_ecs::system::Resource;
use bevy_mod_scripting::core::event::IntoCallbackLabel;
use bevy_mod_scripting::core::script::ScriptId;
use bevy_mod_scripting::core::IntoScriptPluginParams;
use bevy_reflect::Reflect;
use indexmap::IndexSet;
use std::marker::PhantomData;

/// Type-parameterized script storage for tracking active scripts
#[derive(Resource, Default, Reflect)]
pub struct KomoToolScriptStore<P, L>
/*
where
    P: IntoScriptPluginParams
        + ScriptFunctionChecker
        + Send
        + Sync
        + 'static
        + std::default::Default,
    L: IntoCallbackLabel + Send + Sync + 'static + std::default::Default,

 */
{
    #[reflect(ignore)]
    _phantom: PhantomData<(L, P)>,
}
```
this compiles:
```Rust
use super::ScriptFunctionChecker;
use bevy_ecs::system::Resource;
use bevy_mod_scripting::core::event::IntoCallbackLabel;
use bevy_mod_scripting::core::script::ScriptId;
use bevy_mod_scripting::core::IntoScriptPluginParams;
use bevy_reflect::Reflect;
use indexmap::IndexSet;
use std::marker::PhantomData;

/// Type-parameterized script storage for tracking active scripts
#[derive(Resource, Default, Reflect)]
pub struct KomoToolScriptStore<P>
where
    P: IntoScriptPluginParams
        + ScriptFunctionChecker
        + Send
        + Sync
        + 'static
        + std::default::Default,
    //L: IntoCallbackLabel + Send + Sync + 'static + std::default::Default,
{
    /// Set of active script identifiers
    #[reflect(ignore)]
    pub scripts: IndexSet<ScriptId>,
    #[reflect(ignore)]
    _phantom: PhantomData<P>,
}
```
```Rust
use super::ScriptFunctionChecker;
use bevy_ecs::system::Resource;
use bevy_mod_scripting::core::event::IntoCallbackLabel;
use bevy_mod_scripting::core::script::ScriptId;
use bevy_mod_scripting::core::IntoScriptPluginParams;
use bevy_reflect::Reflect;
use indexmap::IndexSet;
use std::marker::PhantomData;

/// Type-parameterized script storage for tracking active scripts
#[derive(Resource, Default, Reflect)]
pub struct KomoToolScriptStore<L>
where
    /*
    P: IntoScriptPluginParams
        + ScriptFunctionChecker
        + Send
        + Sync
        + 'static
        + std::default::Default,

     */
    L: IntoCallbackLabel + Send + Sync + 'static + std::default::Default,
{
    /// Set of active script identifiers
    #[reflect(ignore)]
    pub scripts: IndexSet<ScriptId>,
    #[reflect(ignore)]
    _phantom: PhantomData<L>,
}
```

this also compiles 

this is not compiling as well:
```Rust
use super::ScriptFunctionChecker;
use bevy_ecs::system::Resource;
use bevy_mod_scripting::core::event::IntoCallbackLabel;
use bevy_mod_scripting::core::script::ScriptId;
use bevy_mod_scripting::core::IntoScriptPluginParams;
use bevy_reflect::Reflect;
use indexmap::IndexSet;
use std::marker::PhantomData;

/// Type-parameterized script storage for tracking active scripts
#[derive(Resource, Default, Reflect)]
pub struct KomoToolScriptStore<P, L>
/*
where
    P: IntoScriptPluginParams
        + ScriptFunctionChecker
        + Send
        + Sync
        + 'static
        + std::default::Default
        + Clone,
    L: IntoCallbackLabel + Send + Sync + 'static + std::default::Default + Clone,

 */
{
    #[reflect(ignore)]
    _phantom: PhantomData<(L, P)>,
}
```

there is a discussion in reflection-dev for more information

--- a/crates/bevy_reflect/derive/src/string_expr.rs
+++ b/crates/bevy_reflect/derive/src/string_expr.rs
@@ -80,7 +80,7 @@ impl StringExpr {
         let owned = self.into_owned();
         let borrowed = other.into_borrowed();
         Self::Owned(quote! {
-            #owned + #borrowed
+            ::core::ops::Add::<&str>::add(#owned, #borrowed)
         })
     }
 }

--- a/crates/bevy_reflect/src/lib.rs
+++ b/crates/bevy_reflect/src/lib.rs
@@ -988,6 +988,41 @@ mod tests {
         assert_eq!(values, vec![1]);
     }
 
+    /// This test ensures that we are able to reflect generic types with one or more type parameters.
+    ///
+    /// When there is an `Add` implementation for `String`, the compiler isn't able to infer the correct
+    /// type to deref to.
+    /// If we don't append the strings in the `TypePath` derive correctly (i.e. explicitly specifying the type),
+    /// we'll get a compilation error saying that "`&String` cannot be added to `String`".
+    ///
+    /// So this test just ensures that we do do that correctly.
+    ///
+    /// This problem is a known issue and is unexpectedly expected behavior:
+    /// - <https://github.com/rust-lang/rust/issues/77143>
+    /// - <https://github.com/bodil/smartstring/issues/7>
+    /// - <https://github.com/pola-rs/polars/issues/14666>
+    #[test]
+    fn should_reflect_generic() {
+        struct FakeString {}
+
+        // This implementation confuses the compiler when trying to add a `&String` to a `String`
+        impl core::ops::Add<FakeString> for String {
+            type Output = Self;
+            fn add(self, _rhs: FakeString) -> Self::Output {
+                unreachable!()
+            }
+        }
+
+        #[derive(Reflect)]
+        struct Foo<A>(A);
+
+        #[derive(Reflect)]
+        struct Bar<A, B>(A, B);
+
+        #[derive(Reflect)]
+        struct Baz<A, B, C>(A, B, C);
+    }
+
     #[test]
     fn should_reflect_clone() {
         // Struct

